backend/src/schema/resolvers/images/images.js

```diff
@@ -0,0 +1,121 @@
+import path from 'path'
+import { v4 as uuid } from 'uuid'
+import slug from 'slug'
+import { existsSync, unlinkSync, createWriteStream } from 'fs'
+import { getDriver } from '../../../db/neo4j'
+import { UserInputError } from 'apollo-server'
+
+// const widths = [34, 160, 320, 640, 1024]
+
+export async function deleteImage(resource, relationshipType, opts = {}) {
+  sanitizeRelationshipType(relationshipType)
+  const { transaction, deleteCallback } = opts
+  if (!transaction) return wrapTransaction(deleteImage, [resource, relationshipType], opts)
+  const txResult = await transaction.run(
+    `
+    MATCH (resource {id: $resource.id})-[rel:${relationshipType}]->(image:Image)
+    WITH image, image {.*} as imageProps
+    DETACH DELETE image
+    RETURN imageProps
+  `,
+    { resource },
+  )
+  const [image] = txResult.records.map(record => record.get('imageProps'))
+  // This behaviour differs from `mergeImage`. If you call `mergeImage`
+  // with metadata for an image that does not exist, it's an indicator
+  // of an error (so throw an error). If we bulk delete an image, it
+  // could very well be that there is no image for the resource.
+  if (image) deleteImageFile(image, deleteCallback)
+  return image
+}
+
+export async function mergeImage(resource, relationshipType, imageInput, opts = {}) {
+  if (typeof imageInput === 'undefined') return
+  if (imageInput === null) return deleteImage(resource, relationshipType, opts)
+  sanitizeRelationshipType(relationshipType)
+  const { transaction, uploadCallback, deleteCallback } = opts
+  if (!transaction)
+    return wrapTransaction(mergeImage, [resource, relationshipType, imageInput], opts)
+
+  let txResult
+  txResult = await transaction.run(
+    `
+    MATCH (resource {id: $resource.id})-[:${relationshipType}]->(image:Image)
+    RETURN image {.*}
+    `,
+    { resource },
+  )
+  const [existingImage] = txResult.records.map(record => record.get('image'))
+  const { upload } = imageInput
+  if (!(existingImage || upload)) throw new UserInputError('Cannot find image for given resource')
+  if (existingImage && upload) deleteImageFile(existingImage, deleteCallback)
+  const url = await uploadImageFile(upload, uploadCallback)
+  const { alt, sensitive, aspectRatio } = imageInput
+  const image = { alt, sensitive, aspectRatio, url }
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-02-25T15:32:37Z" title="Tuesday, February 25th 2020, 4:32:37 pm +01:00">Feb 25, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,120 @@
+import path from 'path'
+import uuid from 'uuid/v4'
+import slug from 'slug'
+import { existsSync, unlinkSync, createWriteStream } from 'fs'
+import { getDriver } from '../../../db/neo4j'
+import { UserInputError } from 'apollo-server'
+
+// const widths = [34, 160, 320, 640, 1024]
+
+export async function deleteImage(resource, relationshipType, opts = {}) {
+  sanitizeRelationshipType(relationshipType)
+  const { transaction, deleteCallback } = opts
+  if (!transaction) return wrapTransaction(deleteImage, [resource, relationshipType], opts)
+  const txResult = await transaction.run(
+    `
+    MATCH (resource {id: $resource.id})-[rel:${relationshipType}]->(image:Image)
+    WITH image, image {.*} as imageProps
+    DETACH DELETE image
+    RETURN imageProps
+  `,
+    { resource },
+  )
+  const [image] = txResult.records.map(record => record.get('imageProps'))
+  // This behaviour differs from `mergeImage`. If you call `mergeImage`
+  // with metadata for an image that does not exist, it's an indicator
+  // of an error (so throw an error). If we bulk delete an image, it
+  // could very well be that there is no image for the resource.
+  if (image) deleteImageFile(image, deleteCallback)
+  return image
+}
+
+export async function mergeImage(resource, relationshipType, imageInput, opts = {}) {
+  if (typeof imageInput === 'undefined') return
+  if (imageInput === null) return deleteImage(resource, relationshipType, opts)
+  sanitizeRelationshipType(relationshipType)
+  const { transaction, uploadCallback, deleteCallback } = opts
+  if (!transaction)
+    return wrapTransaction(mergeImage, [resource, relationshipType, imageInput], opts)
+
+  let txResult
+  txResult = await transaction.run(
+    `
+    MATCH (resource {id: $resource.id})-[:${relationshipType}]->(image:Image)
+    RETURN image {.*}
+    `,
+    { resource },
+  )
+  const [existingImage] = txResult.records.map(record => record.get('image'))
+  const { upload } = imageInput
+  if (!(existingImage || upload)) throw new UserInputError('Cannot find image for given resource')
+  if (existingImage && upload) deleteImageFile(existingImage, deleteCallback)
+  const url = await uploadImageFile(upload, uploadCallback)
+  const { alt, sensitive, aspectRatio } = imageInput
+  const image = { alt, sensitive, aspectRatio, url }
```


maybe I don't understand what you are doing here, but to me it looks like you are destructuring these variables and then saving them to an object, isn't this the same as `const image = { ...imageInput, url }`??

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2020-03-02T10:39:15Z" title="Monday, March 2nd 2020, 11:39:15 am +01:00">Mar 2, 2020</time>_

---


`imageInput` contains the `upload`. You don't want that to be part of `image`.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-03-11T19:25:25Z" title="Wednesday, March 11th 2020, 8:25:25 pm +01:00">Mar 11, 2020</time>_

---


thanks for the explanation... I had forgotten about that